File: app/cli/swift/Tests/IntegrationTests/Harness.swift
// Boots a real `pocketskynet` process per suite, modeled on
// `app/server/tests/common/harness.rs`:
//
// * ephemeral ports (TCP and UDP live in different namespaces, so each is
//   probed in its own),
// * boots serialized under a lock — "ask the OS for a free port, close it,
//   hand it to a child" has a race window, and the loser of a bind race gets
//   its `/api/health` probe answered by the *winner*, so after health goes
//   green we re-check that our own child is still alive,
// * a temp data directory per server, removed on teardown,
// * teardown in class tearDown (runs even when tests fail) plus deinit.
//
// The server binary is found by walking up from this file for
// `app/target/{release,debug}/pocketskynet` — robust to worktrees and
// absorbed-submodule layouts, where the target dir lives above the checkout
// that holds the sources. `POCKETSKYNET_BIN` overrides. If no binary exists
// anywhere, one `cargo build --release` is attempted in the nearest `app/`.

import Foundation
import PocketSkynetClient
import XCTest

final class TestServer {
    struct BootFailure: Error, CustomStringConvertible {
        let message: String
        var description: String { message }
    }

    static let jwtSecret = "pocketskynet-integration-test-secret-0123456789abcdef"
    private static let bootLock = NSLock()
    private static let bootTimeout: TimeInterval = 30

    let process: Process
    let port: UInt16
    let http3Port: UInt16?
    let dataDir: URL
    let tls: Bool
    /// `http(s)://127.0.0.1:<port>` — the TCP listener.
    let baseURL: String

    /// The QUIC endpoint's base URL. HTTP/3 is always TLS.
    var http3URL: String {
        guard let http3Port else { preconditionFailure("this server has no HTTP/3 listener") }
        return "https://127.0.0.1:\(http3Port)"
    }

    var logPath: URL { dataDir.appendingPathComponent("server.log") }

    var serverLog: String {
        (try? String(contentsOf: logPath, encoding: .utf8)) ?? ""
    }

    private init(process: Process, port: UInt16, http3Port: UInt16?, dataDir: URL, tls: Bool) {
        self.process = process
        self.port = port
        self.http3Port = http3Port
        self.dataDir = dataDir
        self.tls = tls
        self.baseURL = "\(tls ? "https" : "http")://127.0.0.1:\(port)"
    }

    // MARK: Boot

    /// Start a server and block until `/api/health` answers 200 from *our*
    /// child. Retries the whole boot on a lost bind race.
    static func start(tls: Bool = false, http3: Bool = false) -> TestServer {
        var lastError = ""
        for _ in 0..<5 {
            switch tryStart(tls: tls, http3: http3) {
            case .success(let server): return server
            case .failure(let error): lastError = error.message
            }
        }
        fatalError("could not start pocketskynet after 5 attempts: \(lastError)")
    }

    private static func tryStart(tls: Bool, http3: Bool) -> Result<TestServer, BootFailure> {
        bootLock.lock()
        defer { bootLock.unlock() }

        let binary = locateBinary()
        let port = freeTCPPort()
        let quicPort = http3 ? freeUDPPort() : nil

        let dataDir = FileManager.default.temporaryDirectory
            .appendingPathComponent("ps-swift-it-\(ProcessInfo.processInfo.processIdentifier)-\(UUID().uuidString)")
        let staticDir = dataDir.appendingPathComponent("static")
        do {
            try FileManager.default.createDirectory(at: staticDir, withIntermediateDirectories: true)
        } catch {
            return .failure(BootFailure(message: "mkdir \(staticDir.path): \(error)"))
        }

        var arguments = [
            "--host", "127.0.0.1",
            "--port", String(port),
            "--data-dir", dataDir.path,
            "--static-dir", staticDir.path,
            "--jwt-secret", jwtSecret,
            "--no-rate-limit",
            "--no-payment-verify",
            "--no-mdns",
            "--log", "warn",
        ]
        if tls { arguments.append("--tls") }
        if let quicPort {
            arguments.append(contentsOf: ["--http3", "--http3-port", String(quicPort)])
        }

        let process = Process()
        process.executableURL = binary
        process.arguments = arguments
        process.environment = scrubbedEnvironment()
        FileManager.default.createFile(atPath: dataDir.appendingPathComponent("server.log").path, contents: nil)
        let log = FileHandle(forWritingAtPath: dataDir.appendingPathComponent("server.log").path)
        process.standardOutput = log
        process.standardError = log
        process.standardInput = FileHandle.nullDevice

        do {
            try process.run()
        } catch {
            return .failure(BootFailure(message: "spawn \(binary.path): \(error)"))
        }

        let server = TestServer(process: process, port: port, http3Port: quicPort, dataDir: dataDir, tls: tls)
        if let failure = server.awaitHealth() {
            let tail = server.serverLog.split(separator: "\n").suffix(30).joined(separator: "\n")
            server.stop()
            return .failure(BootFailure(message: "\(failure)\n--- server log (tail) ---\n\(tail)"))
        }
        return .success(server)
    }

    /// A developer's shell must not decide what the suite tests — drop every
    /// PS_*/VITE_* override, and neutralize values baked in at compile time.
    private static func scrubbedEnvironment() -> [String: String] {
        var env = ProcessInfo.processInfo.environment
        for key in Array(env.keys) where key.hasPrefix("PS_") || key.hasPrefix("VITE_") {
            env.removeValue(forKey: key)
        }
        env.removeValue(forKey: "POCKETSKYNET_PATH")
        env["PS_IGNORE_BAKED_ENV"] = "1"
        return env
    }

    /// `nil` when healthy; otherwise a description of what went wrong.
    private func awaitHealth() -> String? {
        let deadline = Date().addingTimeInterval(Self.bootTimeout)
        let url = "\(baseURL)/api/health"
        while Date() < deadline {
            if !process.isRunning {
                return "server exited during boot with status \(process.terminationStatus)"
            }
            if Self.syncGET(url, insecure: tls) == 200 {
                // Somebody answered — make sure it was us. If our child lost
                // the bind race it has already exited.
                if process.isRunning {
                    return nil
                }
                return "another process owns port \(port); our child exited with status \(process.terminationStatus)"
            }
            Thread.sleep(forTimeInterval: 0.02)
        }
        return "/api/health never became ready on port \(port)"
    }

    /// Blocking GET so the harness can run inside synchronous XCTest setUp.
    static func syncGET(_ url: String, insecure: Bool) -> Int? {
        let transport = URLSessionTransport(insecure: insecure, http3: false, timeout: 2)
        let semaphore = DispatchSemaphore(value: 0)
        var status: Int?
        Task.detached {
            defer { semaphore.signal() }
            guard let target = URL(string: url) else { return }
            status = try? await transport.send(method: "GET", url: target, headers: [:], body: nil).status
        }
        semaphore.wait()
        return status
    }

    // MARK: Teardown

    func stop() {
        if process.isRunning {
            kill(process.processIdentifier, SIGKILL)
            process.waitUntilExit()
        }
        try? FileManager.default.removeItem(at: dataDir)
    }

    deinit {
        if process.isRunning {
            kill(process.processIdentifier, SIGKILL)
        }
    }

    // MARK: Binary location

    private static var cachedBinary: URL?

    static func locateBinary() -> URL {
        if let cachedBinary { return cachedBinary }

        if let override = ProcessInfo.processInfo.environment["POCKETSKYNET_BIN"],
           FileManager.default.isExecutableFile(atPath: override) {
            let url = URL(fileURLWithPath: override)
            cachedBinary = url
            return url
        }

        if let found = searchAncestors() {
            cachedBinary = found
            return found
        }

        // No binary anywhere above us — build it once.
        if let appDir = nearestAppDir() {
            fputs("pocketskynet binary not found; running `cargo build --release` in \(appDir.path) (once)…\n", stderr)
            let build = Process()
            build.executableURL = URL(fileURLWithPath: "/usr/bin/env")
            build.arguments = ["cargo", "build", "--release", "--bin", "pocketskynet"]
            build.currentDirectoryURL = appDir
            do {
                try build.run()
                build.waitUntilExit()
            } catch {
                fatalError("could not run cargo: \(error)")
            }
            if let found = searchAncestors() {
                cachedBinary = found
                return found
            }
        }
        fatalError("""
            No pocketskynet server binary. Build it first:
                cd app && cargo build --release --bin pocketskynet
            or point POCKETSKYNET_BIN at one.
            """)
    }

    /// Walk up from this file, checking `app/target/{release,debug}/pocketskynet`
    /// under every ancestor. When both profiles exist, the newer build wins.
    private static func searchAncestors() -> URL? {
        var dir = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        for _ in 0..<12 {
            let candidates = ["release", "debug"]
                .map { dir.appendingPathComponent("app/target/\($0)/pocketskynet") }
                .filter { FileManager.default.isExecutableFile(atPath: $0.path) }
            if !candidates.isEmpty {
                return candidates.max { modificationDate($0) < modificationDate($1) }
            }
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path { break }
            dir = parent
        }
        return nil
    }

    private static func modificationDate(_ url: URL) -> Date {
        (try? FileManager.default.attributesOfItem(atPath: url.path)[.modificationDate] as? Date)
            .flatMap { $0 } ?? .distantPast
    }

    /// The nearest ancestor `app/` directory holding a Cargo.toml.
    private static func nearestAppDir() -> URL? {
        var dir = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        for _ in 0..<12 {
            let candidate = dir.appendingPathComponent("app/Cargo.toml")
            if FileManager.default.fileExists(atPath: candidate.path) {
                return dir.appendingPathComponent("app")
            }
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path { break }
            dir = parent
        }
        return nil
    }

    // MARK: Ports

    private static func freePort(socketType: Int32) -> UInt16 {
        let fd = socket(AF_INET, socketType, 0)
        precondition(fd >= 0, "socket() failed")
        defer { close(fd) }

        var address = sockaddr_in()
        address.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        address.sin_family = sa_family_t(AF_INET)
        address.sin_port = 0
        address.sin_addr = in_addr(s_addr: inet_addr("127.0.0.1"))

        let bound = withUnsafePointer(to: &address) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                bind(fd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        precondition(bound == 0, "bind() failed: \(String(cString: strerror(errno)))")

        var assigned = sockaddr_in()
        var length = socklen_t(MemoryLayout<sockaddr_in>.size)
        let named = withUnsafeMutablePointer(to: &assigned) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                getsockname(fd, $0, &length)
            }
        }
        precondition(named == 0, "getsockname() failed")
        return UInt16(bigEndian: assigned.sin_port)
    }

    static func freeTCPPort() -> UInt16 { freePort(socketType: SOCK_STREAM) }

    /// TCP and UDP port numbers live in different namespaces; a QUIC listener
    /// needs a port probed as UDP.
    static func freeUDPPort() -> UInt16 { freePort(socketType: SOCK_DGRAM) }
}

// MARK: - Shared test conveniences

extension TestServer {
    /// A client for the TCP listener. TLS servers use self-signed certs, so
    /// their clients run with the trust override.
    func client() -> SkynetClient {
        SkynetClient(baseURL: URL(string: baseURL)!, insecure: tls, http3: false)
    }

    /// A client for the QUIC listener.
    func http3Client() -> SkynetClient {
        SkynetClient(baseURL: URL(string: http3URL)!, insecure: true, http3: true)
    }

    /// A fresh wallet logged in through the whole challenge flow.
    @discardableResult
    func loginFreshUser(username: String? = nil) async throws -> (SkynetClient, EthereumWallet, LoginResponse) {
        let wallet = try EthereumWallet.random()
        let client = client()
        let response = try await client.login(wallet: wallet, username: username)
        return (client, wallet, response)
    }
}

File: app/cli/typescript/test/helpers/harness.ts
/**
 * Boots a real `pocketskynet` process per test file, modeled on
 * `app/server/tests/common/harness.rs`.
 *
 * Every server gets its own ephemeral port and temp data directory. Ports are
 * picked by binding to 0 and releasing, which leaves a race window: another
 * process can win the bind, and our child's `/api/health` probe would then be
 * answered by the winner. So after health answers we check the child is still
 * alive ("is our child alive"), and retry the whole boot when it is not.
 *
 * Teardown kills the child and removes the data dir; a process-exit hook
 * backstops leaked children even when a test file crashes.
 */

import { execFileSync, spawn, ChildProcess } from "node:child_process";
import { randomBytes } from "node:crypto";
import { createSocket } from "node:dgram";
import { existsSync, mkdirSync, openSync, readFileSync, rmSync, statSync } from "node:fs";
import { createServer } from "node:net";
import { tmpdir } from "node:os";
import { isAbsolute, join } from "node:path";
import { setTimeout as sleep } from "node:timers/promises";
import { Agent, fetch as undiciFetch } from "undici";
import { normalizePrivateKey } from "../../src/wallet.js";
import { appRoot } from "./vectors.js";

/** Handed to the server with `--jwt-secret` so tests can mint and tamper. */
export const JWT_SECRET = "pocketskynet-ts-integration-test-secret-0123456789abcdef";

const BOOT_TIMEOUT_MS = 30_000;
let counter = 0;

const liveChildren = new Set<ChildProcess>();
process.on("exit", () => {
  for (const child of liveChildren) {
    try {
      child.kill("SIGKILL");
    } catch {
      /* already gone */
    }
  }
});

/** Locate the pocketskynet binary; build it once with cargo when absent. */
export function resolveServerBinary(): string {
  const fromEnv = process.env["POCKETSKYNET_BIN"];
  if (fromEnv !== undefined && fromEnv.length > 0) {
    if (!existsSync(fromEnv)) {
      throw new Error(`POCKETSKYNET_BIN points at a missing file: ${fromEnv}`);
    }
    return fromEnv;
  }

  const roots = [appRoot()];
  // The checkout may be a git worktree of a normal repo (common dir is
  // `<main>/.git`) or of an absorbed submodule (common dir is
  // `<super>/.git/modules/<name>`, with the main checkout named by its
  // `core.worktree`). Either way, target/ lives with the main checkout.
  try {
    const commonDir = execFileSync(
      "git",
      ["rev-parse", "--path-format=absolute", "--git-common-dir"],
      { cwd: appRoot(), encoding: "utf8" },
    ).trim();
    const besideGitDir = join(commonDir, "..", "app");
    if (existsSync(besideGitDir)) roots.push(besideGitDir);
    try {
      const coreWorktree = execFileSync(
        "git",
        ["config", "--file", join(commonDir, "config"), "core.worktree"],
        { encoding: "utf8" },
      ).trim();
      if (coreWorktree.length > 0) {
        // core.worktree may be absolute or relative to the git dir. join()
        // silently concatenates an absolute second segment, so resolve the
        // absolute case explicitly rather than mis-joining it.
        const worktreeRoot = isAbsolute(coreWorktree)
          ? coreWorktree
          : join(commonDir, coreWorktree);
        const mainRepoApp = join(worktreeRoot, "app");
        if (existsSync(mainRepoApp)) roots.push(mainRepoApp);
      }
    } catch {
      /* no core.worktree — a normal repository */
    }
  } catch {
    /* not a git checkout — fine */
  }

  const candidates: string[] = [];
  for (const root of roots) {
    candidates.push(join(root, "target", "release", "pocketskynet"));
    candidates.push(join(root, "target", "debug", "pocketskynet"));
  }
  const existing = candidates.filter((p) => existsSync(p));
  if (existing.length > 0) {
    // Prefer the most recently built one.
    existing.sort((a, b) => statSync(b).mtimeMs - statSync(a).mtimeMs);
    return existing[0]!;
  }

  // Build once via cargo, then re-check.
  try {
    execFileSync("cargo", ["build", "-p", "pocketskynet"], {
      cwd: appRoot(),
      stdio: "inherit",
      timeout: 15 * 60 * 1000,
    });
  } catch (err) {
    throw new Error(
      `no pocketskynet binary found (looked at ${candidates.join(", ")}) and ` +
        `'cargo build -p pocketskynet' failed: ${err instanceof Error ? err.message : err}. ` +
        "Build the server or set POCKETSKYNET_BIN.",
    );
  }
  const built = join(appRoot(), "target", "debug", "pocketskynet");
  if (!existsSync(built)) {
    throw new Error(`cargo build succeeded but ${built} does not exist`);
  }
  return built;
}

function freeTcpPort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const srv = createServer();
    srv.listen(0, "127.0.0.1", () => {
      const address = srv.address();
      if (address === null || typeof address === "string") {
        srv.close();
        reject(new Error("no TCP address"));
        return;
      }
      const port = address.port;
      srv.close(() => resolve(port));
    });
    srv.on("error", reject);
  });
}

/** TCP and UDP port numbers live in different namespaces — probe UDP itself. */
function freeUdpPort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const sock = createSocket("udp4");
    sock.bind(0, "127.0.0.1", () => {
      const port = sock.address().port;
      sock.close(() => resolve(port));
    });
    sock.on("error", reject);
  });
}

function uniqueDir(): string {
  counter += 1;
  return join(tmpdir(), `ps-ts-it-${process.pid}-${Date.now()}-${counter}`);
}

/** Environment for the child: PS_ and VITE_ vars scrubbed, baked env ignored. */
function childEnv(extra: Record<string, string>): NodeJS.ProcessEnv {
  const env: NodeJS.ProcessEnv = { ...process.env };
  for (const key of Object.keys(env)) {
    if (key.startsWith("PS_") || key.startsWith("VITE_")) delete env[key];
  }
  delete env["POCKETSKYNET_PATH"];
  env["PS_IGNORE_BAKED_ENV"] = "1";
  for (const [key, value] of Object.entries(extra)) env[key] = value;
  return env;
}

export interface StartOptions {
  tls?: boolean;
  http3?: boolean;
  extraArgs?: string[];
  env?: Record<string, string>;
}

export class TestServer {
  private constructor(
    private readonly child: ChildProcess,
    readonly port: number,
    readonly http3Port: number | undefined,
    readonly dataDir: string,
    readonly baseUrl: string,
  ) {}

  get isTls(): boolean {
    return this.baseUrl.startsWith("https://");
  }

  url(path: string): string {
    return `${this.baseUrl}${path}`;
  }

  /** The CA this server generated (TLS and/or HTTP/3 servers). */
  caPath(): string {
    return join(this.dataDir, "tls", "ca.crt");
  }

  caPem(): string {
    return readFileSync(this.caPath(), "utf8");
  }

  http3Url(path: string): string {
    if (this.http3Port === undefined) throw new Error("server has no HTTP/3 listener");
    return `https://127.0.0.1:${this.http3Port}${path}`;
  }

  serverLog(): string {
    try {
      return readFileSync(join(this.dataDir, "server.log"), "utf8");
    } catch {
      return "";
    }
  }

  /** Kill the child and delete the data dir. Safe to call twice. */
  async stop(): Promise<void> {
    try {
      this.child.kill("SIGKILL");
    } catch {
      /* already gone */
    }
    liveChildren.delete(this.child);
    // Give the OS a beat to release the process before removing its cwd data.
    await sleep(30);
    try {
      rmSync(this.dataDir, { recursive: true, force: true });
    } catch {
      /* best effort */
    }
  }

  static async start(opts: StartOptions = {}): Promise<TestServer> {
    let lastErr = "";
    for (let attempt = 0; attempt < 5; attempt++) {
      try {
        return await TestServer.tryStart(opts);
      } catch (err) {
        lastErr = err instanceof Error ? err.message : String(err);
      }
    }
    throw new Error(`could not start pocketskynet after 5 attempts: ${lastErr}`);
  }

  private static async tryStart(opts: StartOptions): Promise<TestServer> {
    const binary = resolveServerBinary();
    const port = await freeTcpPort();
    const http3Port = opts.http3 ? await freeUdpPort() : undefined;
    const dataDir = uniqueDir();
    const staticDir = join(dataDir, "static");
    mkdirSync(staticDir, { recursive: true });

    const logPath = join(dataDir, "server.log");
    const logFd = openSync(logPath, "w");

    const args = [
      "--host",
      "127.0.0.1",
      "--port",
      String(port),
      "--data-dir",
      dataDir,
      "--static-dir",
      staticDir,
      "--jwt-secret",
      JWT_SECRET,
      "--no-rate-limit",
      "--no-payment-verify",
      "--no-mdns",
      "--log",
      "warn",
    ];
    if (opts.tls) args.push("--tls");
    if (opts.http3) args.push("--http3", "--http3-port", String(http3Port));
    if (opts.extraArgs) args.push(...opts.extraArgs);

    const child = spawn(binary, args, {
      env: childEnv(opts.env ?? {}),
      stdio: ["ignore", logFd, logFd],
    });
    liveChildren.add(child);

    const scheme = opts.tls ? "https" : "http";
    const baseUrl = `${scheme}://127.0.0.1:${port}`;
    const server = new TestServer(child, port, http3Port, dataDir, baseUrl);

    try {
      await server.awaitHealth();
      if (opts.http3 || opts.tls) await server.awaitCa();
      return server;
    } catch (err) {
      const log = server.serverLog().split("\n").slice(-30).join("\n");
      await server.stop();
      throw new Error(
        `${err instanceof Error ? err.message : err}\n--- server log (tail) ---\n${log}`,
      );
    }
  }

  private async awaitHealth(): Promise<void> {
    const deadline = Date.now() + BOOT_TIMEOUT_MS;
    // Over TLS, probe with verification off — the real tests then verify
    // against the generated CA explicitly.
    const dispatcher = this.isTls
      ? new Agent({ connect: { rejectUnauthorized: false } })
      : undefined;
    try {
      while (Date.now() < deadline) {
        if (this.child.exitCode !== null) {
          throw new Error(`server exited during boot with code ${this.child.exitCode}`);
        }
        try {
          const response = await undiciFetch(this.url("/api/health"), {
            ...(dispatcher !== undefined ? { dispatcher } : {}),
            signal: AbortSignal.timeout(2_000),
          });
          if (response.ok) {
            await response.body?.cancel();
            // Somebody answered — make sure it was our child. A child that
            // lost the bind race has already exited, and this 200 came from
            // the winner.
            if (this.child.exitCode !== null) {
              throw new Error(
                `another process owns port ${this.port}; our child exited with ${this.child.exitCode}`,
              );
            }
            return;
          }
          await response.body?.cancel();
        } catch (err) {
          if (err instanceof Error && /owns port|exited during boot/.test(err.message)) throw err;
          // Connection refused while booting — keep polling.
        }
        await sleep(25);
      }
      throw new Error(`/api/health never became ready on port ${this.port}`);
    } finally {
      await dispatcher?.close();
    }
  }

  /** `/api/health` answering does not prove the CA file is flushed yet. */
  private async awaitCa(): Promise<void> {
    const deadline = Date.now() + 10_000;
    while (!existsSync(this.caPath()) && Date.now() < deadline) {
      await sleep(20);
    }
    if (!existsSync(this.caPath())) {
      throw new Error(`server never wrote its CA at ${this.caPath()}`);
    }
  }
}

/** A fresh random private key (retrying the astronomically unlikely invalid). */
export function randomPrivateKeyHex(): string {
  for (;;) {
    const candidate = `0x${randomBytes(32).toString("hex")}`;
    try {
      normalizePrivateKey(candidate); // throws for 0 or >= n
      return candidate;
    } catch {
      /* try again */
    }
  }
}
